### PR TITLE
Dispatch documents with common ID in order

### PR DESCRIPTION
**Pull request by mpolden**
@jonmv

### CODE REVIEW

File: client/go/internal/vespa/document/dispatcher.go [hunk 1]
```diff
@@ -5,59 +5,125 @@ import (
...
-	d.readPending()
+	d.start()
 	return d
 }
 
-func (d *Dispatcher) readPending() {
-	for i := 0; i < d.concurrencyLevel; i++ {
+func (d *Dispatcher) dispatchAll(g *documentGroup) int {
+	g.mu.Lock()
+	defer g.mu.Unlock()
```
Comment (jonmv) [inline client/go/internal/vespa/document/dispatcher.go:55]: Hmm, won't this block the client thread if it tries to add another document with this ID, until all operations for that ID have completed? I guess it's not likely to be a problem ... but is the locking necessary?

File: client/go/internal/vespa/document/dispatcher.go [hunk 2]
```diff
@@ -5,59 +5,125 @@ import (
...
+	failCount := len(g.operations)
+	for i := 0; !g.failed && i < len(g.operations); i++ {
```
Comment (jonmv) [inline client/go/internal/vespa/document/dispatcher.go:57]: I'm not sure what we'd like to do when a single operation against a particular document ID fails. In the Java feed client, we just keep going.

**Review by jonmv — APPROVED** [review]
It'll be interesting to see the results of a performance test, but I guess that's quite a while away? :)

File: client/go/internal/vespa/document/dispatcher.go [hunk 3]
```diff
@@ -5,59 +5,125 @@ import (
...
+	for i := 0; i < d.workers; i++ {
 		d.wg.Add(1)
-		go func(n int) {
+		go func() {
 			defer d.wg.Done()
-			for doc := range d.pending {
-				d.feeder.Send(doc)
+			for id := range d.ready {
+				d.mu.RLock()
+				group := d.inflight[id.String()]
```
[the remark this answers is not in the document]
Comment (mpolden) [reply client/go/internal/vespa/document/dispatcher.go:88]: Not currently. Can look into that if memory use becomes a problem with really large document batches.

File: client/go/internal/vespa/document/dispatcher.go (hunk 2 above)
Comment (mpolden) [reply client/go/internal/vespa/document/dispatcher.go:57]: I'm unsure what is right as well. I can change it to match vespa-feed-client behavior in later PR.

**Pull request merged**